src/libsyntax/parse/parser/pat.rs

```diff
@@ -100,40 +155,31 @@ impl<'a> Parser<'a> {
         Ok(())
     }
 
-    /// Parses a pattern, that may be a or-pattern (e.g. `Some(Foo | Bar)`).
-    fn parse_pat_with_or(&mut self, expected: Option<&'static str>) -> PResult<'a, P<Pat>> {
-        // Parse the first pattern.
-        let first_pat = self.parse_pat(expected)?;
-
-        // If the next token is not a `|`, this is not an or-pattern and
-        // we should exit here.
-        if !self.check(&token::BinOp(token::Or)) {
-            return Ok(first_pat)
-        }
-
-        let lo = first_pat.span;
-
-        let mut pats = vec![first_pat];
+    /// Recursive possibly-or-pattern parser with recovery for an erroneous leading `|`.
+    /// See `parse_pat_with_or` for details on parsing or-patterns.
+    fn parse_pat_with_or_inner(&mut self, expected: Expected) -> PResult<'a, P<Pat>> {
+        // Recover if `|` or `||` is here.
+        // The user is thinking that a leading `|` is allowed in this position.
+        if let token::BinOp(token::Or) | token::OrOr = self.token.kind {
```

Comment:
All this could be a separate `recover_bad_leading_or` method.

Reply:
Good point; will fix. 👍